tools/preproc/asm_file.cpp

```diff
@@ -527,6 +507,63 @@ void AsmFile::OutputLine()
     }
 }
 
+// parses an assumed C `enum`. Returns false if `enum { ...` is not matched
+bool AsmFile::ParseEnum()
+{
+    long fallbackPosition = m_pos;
+    std::string headerFilename = "";
+    long currentHeaderLine = SkipWhitespaceAndEol();
+    std::string enumName = ReadIdentifier();
+    currentHeaderLine += SkipWhitespaceAndEol();
+    long enumCounter = 0;
+    long symbolCount = 0;
+
+    if (m_buffer[m_pos] != '{') // assume assembly macro, otherwise assume enum and report errors accordingly
+    {
+        m_pos = fallbackPosition - 4;
+        return false;
+    }
+
+    currentHeaderLine += FindLastLineNumber(headerFilename);
+    m_pos++;
+    for (;;)
+    {
+        currentHeaderLine += SkipWhitespaceAndEol();
+        std::string currentIdentName = ReadIdentifier();
+        if (currentIdentName.empty() && symbolCount == 0)
+        {
+            RaiseError("%s:%ld: empty enum is invalid", headerFilename.c_str(), currentHeaderLine);
+        }
+        std::printf("# %ld \"%s\"\n", currentHeaderLine, headerFilename.c_str());
+        currentHeaderLine += SkipWhitespaceAndEol();
+        if (m_buffer[m_pos] == '=')
+        {
+            m_pos++;
+            currentHeaderLine += SkipWhitespaceAndEol();
+            enumCounter = ReadInteger(headerFilename, currentHeaderLine);
+            currentHeaderLine += SkipWhitespaceAndEol();
+        }
+        
+        std::printf(".equiv %s, %ld\n", currentIdentName.c_str(), enumCounter);
+        enumCounter++;
+        if (m_buffer[m_pos] != ',')
```

Comment:
enums are normally allowed to include a terminal comma (and we do this often for enums where it's likely new entries will be appended), it might be confusing if enums included in assembly require no terminal comma (especially because the error message doesn't explicitly state this).